snuba/datasets/plans/query_plan.py

```diff
@@ -1,46 +1,19 @@
 from __future__ import annotations
 
 from abc import ABC, abstractmethod
-from dataclasses import dataclass
-from typing import Callable, Sequence
+from typing import Generic, TypeVar
 
-from snuba.clickhouse.processors import QueryProcessor
-from snuba.clickhouse.query import Query
-from snuba.clickhouse.sql import SqlQuery
-from snuba.reader import Reader
 from snuba.request import Request
 from snuba.request.request_settings import RequestSettings
 from snuba.web import QueryResult
 
-QueryRunner = Callable[[Query, RequestSettings, Reader[SqlQuery]], QueryResult]
+TStorageQuery = TypeVar("TStorageQuery")
+TQueryRunner = TypeVar("TQueryRunner")
 
 
-@dataclass(frozen=True)
-class ClickhouseQueryPlan:
+class QueryPlanExecutionStrategy(ABC, Generic[TStorageQuery, TQueryRunner]):
```

Comment:
I think I can make the implementation classes generic as well. They do not strictly depend on any clickhouse detail. Still have to figure out a few details.